File: src/SIL.Machine.Shared/Tokenization/LineSegmentTokenizer.cs
using System.Collections.Generic;
using SIL.Machine.Annotations;
using System.Text.RegularExpressions;

namespace SIL.Machine.Tokenization
{
	public class LineSegmentTokenizer : StringTokenizer
	{
		private static readonly Regex NewLineRegex = new Regex("\n|\r\n?");

		public override IEnumerable<Range<int>> Tokenize(string data, Range<int> range)
		{
			int lineStart = 0;
			foreach (Match match in NewLineRegex.Matches(data.Substring(0, range.End), range.Start))
			{
				yield return Range<int>.Create(lineStart, match.Index);
				lineStart = match.Index + match.Length;
			}

			if (lineStart < range.End)
				yield return Range<int>.Create(lineStart, range.End);
		}
	}
}

File: src/SIL.Machine.TS/src/tokenization/line-segment-tokenizer.ts
import { createRange, Range } from '../annotations/range';
import { StringTokenizer } from './string-tokenizer';

export class LineSegmentTokenizer extends StringTokenizer {
  tokenize(data: string, range: Range = createRange(0, data.length)): Range[] {
    const tokens: Range[] = [];
    let lineStart = range.start;
    for (let i = range.start; i < range.end; i++) {
      if (data[i] === '\n' || data[i] === '\r') {
        tokens.push(createRange(lineStart, i));
        if (data[i] === '\r' && i + 1 < range.end && data[i + 1] === '\n') {
          i++;
        }
        lineStart = i + 1;
      }
    }

    if (lineStart < range.end) {
      tokens.push(createRange(lineStart, range.end));
    }

    return tokens;
  }
}
